Replace stream typeahead with dropdown.

whoodes opened the pull request:
This PR addresses issue #11832.
Discussion on [czo](https://chat.zulip.org/#narrow/stream/49-development-help/topic/.2311682/near/715962).

The stream typeahead in the compose box was replaced with a dropdown menu.  This feature helps with on-boarding issues, as well as serving as a replacement for somewhat inconsistent typeahead code.

There are a few points of uncertainty:

* Currently, it is a little hacky how the drop-up is being forced, i.e., by defaulting to the last element of the list if no default value exists.  FF behaves much better than chrome, but after researching, it seems Chrome hands off positioning of the dropdown to the OS. 

* Node tests are passing with branch coverage of 60%, which is pretty sub-par (need a little direction on how to improve this number).

* In terms of styling, unsure how to address the edge case of a long stream name.  Currently, the overflow is hidden, but it looks rather bad as the dropdown arrow overlaps.  Attempts were made to add right padding and trail with an ellipsis, but wasn't able to implement.


**Testing Plan:** <!-- How have you tested? -->

Manual tests were done extensively with regard to both functionality and consistent styling.

Casper tests and node tests were updated.

**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

Below demonstrates clicking on the dropdown from 'new topic' in all messages, along with composing solely with keyboard shortcuts...

![stream-dropdown](https://user-images.githubusercontent.com/39782863/57125795-69a33800-6d26-11e9-86c8-4b5f52f09641.gif)



<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

rishig commented:
cool! A few minor things:
* The dropdown should start blank, rather than with something filled in
* Ideally we would sort the stream names as they are sorted in the left sidebar
* The styling should probably look closer to the stream typeahead in compose (e.g. white background)

whoodes commented:
Sounds good! Although, could you please elaborate on your last bullet point @rishig ?

rishig commented:
E.g. styling more like the following:
![image](https://user-images.githubusercontent.com/890911/54498356-952ba980-48c3-11e9-9bae-e2055e1a2ab7.png)

(Though I think my vision for how this feature would work was a bit different from others, so maybe don't act on it yet.)

whoodes commented:
@timabbott this is ready for review. :)

whoodes commented:
@rishig I have the streams sorted as they appear in the left sidebar, I was wondering if having a "Stream" placeholder like there is currently would be better than an empty string?  With this change it's hidden on the list, but selected in a "stream-less" view. Lmk what you think and I can push the updates.

<img width="369" alt="Screen Shot 2019-03-19 at 11 17 06 AM" src="https://user-images.githubusercontent.com/39782863/54642433-b2ce4f80-4a38-11e9-836e-64ccc07a67f1.png">

rishig commented:
Good question. Is the dropdown opened automatically when you open the compose box?

whoodes commented:
> Good question. Is the dropdown opened automatically when you open the compose box?

Not currently, only focused.

whoodes commented:
Just pushed sorting/inserting logic changes that Rishi spoke of, passing tests locally. Failing at:
```
Mar 20 06:48:10 Get:6 http://ppa.launchpad.net/tabbott/zulip/ubuntu bionic/main amd64 postgresql-10-tsearch-extras amd64 0.4bionic2 [5,360 B]
Too long with no output (exceeded 10m0s)
```

showell commented on static/js/compose.js:
```diff
@@ -1040,6 +1053,31 @@ exports.initialize = function () {
+            var added_stream = streams[streams.length - 1];
+            $("#stream_message_recipient_stream>option").each(function () {
+                if (added_stream.toLowerCase() < $(this).text().toLowerCase()) {
+                    $(this).before('<option value="' + added_stream + '">' + added_stream + '</option>');
```
What happens if you're at the end of the list?

whoodes commented:
Revised the logic per Steve's review and cleaned up the code.  Realized something as I was doing this: Is it pertinent to add logic to deal with updating the dropdown (prior to a refresh) in the event of unsubscribing from a stream?

whoodes commented:
I've squashed the commits down, and reduced complexity of the ```on('focus')``` action.  Any number of streams added or removed now appear in the dropdown.

whoodes commented:
I restructured the commits so they pass all tests individually (the node tests for the typeahead needed to be removed in the second commit).  I replaced the `forEach` loops I was using with `lodash`, and added comments to both code and the commit messages.

timabbott commented:
@rishig are you happy with the UX of this version?

rishig commented:
Sorry for the delay here. As is, it's an improvement for new users, but a regression for experienced users and keyboard users.

Maybe worth chatting about next steps on czo? I think what Tim was imagining is this, plus:
* It becomes scrollable or does something else nice when you have more than 8-10 streams. 
* You can still have a fully-keyboard experience (both in terms of typing letters working, and in terms of arrow keys working)

That seems like a major project though.

rishig commented:
Closing, in light of https://chat.zulip.org/#narrow/stream/49-development-help/topic/dropdown.20stream.20box.

Thanks again for your work on this!
